Type-annotate uvicorn/config.py

br3ndonland opened the pull request:
## Description

This PR will add type annotations and mypy type checking to _uvicorn/config.py_, the associated tests in _tests/test_config.py_, and the related code that is impacted by the new type annotations.

Supersedes the changes to _uvicorn/config.py_ in #992, and closes #1046.

## Changes

### _uvicorn/config.py_

- Add changes from PR #992 (0db3bc2): This commit will bring in changes to _uvicorn/config.py_ added by @Kludex in PR #992, updating for the latest master branch.
- Correct SSL type annotations (eee7fba)
  - `certfile` is a required positional argument when running [`SSLContext.load_cert_chain`](https://docs.python.org/3/library/ssl.html#ssl.SSLContext.load_cert_chain), so annotating as `Optional` (which allows `None`) would not be ideal. Path-like objects are acceptable, so after `from pathlib import Path`, the annotation is `certfile: Union[Path, str]`.
  - `if self.is_ssl and self.ssl_certfile` will help ensure that the `self.ssl_certfile` required positional argument is present.
- Simplify `log_config` type annotation (b2a1548)
- Add type annotation for app (12d323c):
  - app is sometimes a string, as described in _uvicorn/main.py_. In other cases, especially in the tests, app is an ASGI application callable.
  - This commit will add a type annotation for the app argument to address these use cases.
- Simplify if expression in `Config().headers` (b188fd1)
- Allow paths to be used for `Config(env_file)` (d31fddd)
- Add asyncio Protocol types to class `Config` kwargs (9c8e70d): Protocol classes are sometimes used for the app kwarg, such as in _tests/test_config.py_.
- Improve use of `Literal` type for ASGI interfaces (14cb546)
  - This commit will retain the use of [literals](https://mypy.readthedocs.io/en/stable/literal_types.html) to define ASGI protocol versions, but improve and correct the use of literal types.
  - As described in the mypy docs, `Literal["2.0", "3.0"]` is a simpler way to write `Union[Literal["2.0"], Literal["3.0"]]`.
- Add type annotation to `Config` kwargs in _uvicorn/workers.py_ (459544d)
- Add type annotations to _tests/test_config.py_ (c436bba)
- Type-annotate constants in _uvicorn/config.py_ (8d2cd6f): Helps prevent mypy `[has-type]` errors ("Cannot determine type of {object}")

### _uvicorn/importer.py_

Correct type annotations in _uvicorn/importer.py_ (e9e59e9)

The relationship between module imports and calls in _uvicorn/config.py_ was previously unclear. In encode/uvicorn#991, `import_from_string` was annotated as returning a `ModuleType`, but the function is used to return callables and other types. Mypy was raising "module not callable" errors, as reported in #1046.

https://github.com/encode/uvicorn/blob/87da6cf4082cd28306bdb78d7d42552d131cc179/uvicorn/config.py#L317-L325

Current use cases of `import_from_string` include:

- `uvicorn.Config.HTTP_PROTOCOLS`, `uvicorn.Config.WS_PROTOCOLS`: `Type[asyncio.Protocol]` (these are subclasses of asyncio protocols, so they are annotated with `Type[Class]` as explained in the [mypy docs](https://mypy.readthedocs.io/en/stable/kinds_of_types.html#the-type-of-class-objects))
- `uvicorn.Config.LOOP_SETUPS`: `Callable` (each of the loops is a function)
- `uvicorn.Config().loaded_app`: `ASGIApplication` (should be an ASGI application instance, like `Starlette()` or `FastAPI()`)

Ideally, the return type of `import_from_string` would reflect these use cases, but the complex typing will be difficult to maintain. For easier maintenance, the return type will be `Any`, and objects returned by `import_from_string` will be annotated directly.

## Related

encode/uvicorn#990
encode/uvicorn#991
encode/uvicorn#992
encode/uvicorn#998
encode/uvicorn#1046

br3ndonland committed:
Add changes from PR #992

This commit will bring in changes to uvicorn/config.py added by @Kludex
in PR #992, updating for the latest master branch.

br3ndonland committed:
Correct SSL type annotations

encode/uvicorn#992
https://docs.python.org/3/library/ssl.html#ssl.SSLContext.load_cert_chain

- `certfile` is a required positional argument when running
  `SSLContext.load_cert_chain`, so annotating as `Optional` (which
  allows `None`) would not be ideal. Path-like objects are acceptable,
  so after `from pathlib import Path`, the annotation is
  `certfile: Union[Path, str]`.
- `if self.is_ssl and self.ssl_certfile` will help ensure that the
  `self.ssl_certfile` required positional argument is present.

br3ndonland committed:
Simplify log_config type annotation

`Dict[str, Any]` can be simplified to `dict`.

br3ndonland committed:
Add type annotation for app

encode/uvicorn#990
encode/uvicorn#992

app is sometimes a string, as described in uvicorn/main.py. In other
cases, especially in the tests, app is an ASGI application callable.

This commit will add a type annotation for the app argument to address
these use cases.

br3ndonland committed:
Simplify if expression in Config().headers

br3ndonland committed:
Allow paths to be used for Config(env_file)

br3ndonland committed:
Add asyncio Protocol types to class Config kwargs

Protocol classes are sometimes used for the app kwarg, such as in
tests/test_config.py.

br3ndonland committed:
Improve use of Literal type for ASGI interfaces

https://mypy.readthedocs.io/en/stable/literal_types.html

This commit will retain the use of literals to define ASGI protocol
versions, but improve and correct the use of literal types.

As described in the mypy docs, `Literal["2.0", "3.0"]` is a simpler way
to write `Union[Literal["2.0"], Literal["3.0"]]`.

br3ndonland committed:
Add type annotation to Config kwargs in workers.py

https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html

This is just one of those times when mypy needs a little help.

br3ndonland committed:
Add type annotations to test_config.py

PR #978 added logic to convert `reload_dirs` from string to list, as
shown in `test_reload_dir_is_set()`, so the annotation on reload_dirs
will be updated to accept a string.

Kludex commented:
Yeah, the `import_from_string` probably needs to be fixed.

Somehow related: https://github.com/encode/uvicorn/issues/1046

br3ndonland committed:
Type-annotate constants in uvicorn/config.py

Prevents mypy `[has-type]` errors ("Cannot determine type of {object}")

br3ndonland committed:
Correct type annotations in uvicorn/importer.py

encode/uvicorn#1046
encode/uvicorn#1067

The relationship between module imports and calls in uvicorn/config.py
was previously unclear. `import_from_string` was annotated as returning
a `ModuleType`, but was being used as if it were returning callables.
Mypy was raising "module not callable" errors, as reported in #1046.

Current use cases of `import_from_string` include:

- `uvicorn.Config.HTTP_PROTOCOLS`, `uvicorn.Config.WS_PROTOCOLS`:
  `Type[asyncio.Protocol]` (these are subclasses of asyncio protocols,
  so they are annotated with `Type[Class]` as explained in the mypy docs
- `uvicorn.Config.LOOP_SETUPS`: `Callable` (each loop is a function)
- `uvicorn.Config().loaded_app`: `ASGIApplication` (should be an ASGI
  application instance, like `Starlette()` or `FastAPI()`)

Ideally, the return type of `import_from_string` would reflect these use
cases, but the complex typing will be difficult to maintain. For easier
maintenance, the return type will be `Any`, and objects returned by
`import_from_string` will be annotated directly.

br3ndonland commented:
> Yeah, the `import_from_string` probably needs to be fixed.
> 
> Somehow related: #1046

_EDIT:_ I attempted to add a type alias for the return type of `import_from_string`, but there were many complex mypy errors and it was difficult to manage. Rather than trying to be specific with the return type of `import_from_string`, I made the return type `Any`, and shifted the specific type annotations to the objects returned by calls to `import_from_string`. Should be easier to maintain.

Mypy and tests are passing on my fork.

euri10 commented on tests/importer/test_importer.py:
```diff
@@ -3,41 +3,41 @@
-    instance = import_from_string(TemporaryFile)
-    assert instance == TemporaryFile
+    instance = import_from_string(on)
+    assert instance == on
```
not sure if it's the diff or me but I dont understand the change

br3ndonland replied:
Yeah, this was to test the type annotation added in encode/uvicorn#991 (which said `import_from_str` would accept either a module or string). I relaxed the type annotation to just accept `Any`, and reverted the change here.

euri10 commented on tests/test_config.py:
```diff
@@ -64,12 +67,12 @@ def test_config_should_reload_is_set(app, expected_should_reload):
-def test_reload_dir_is_set():
-    config = Config(app=asgi_app, reload=True, reload_dirs="reload_me")
+def test_reload_dir_is_set() -> None:
+    config = Config(app=asgi_app, reload=True, reload_dirs=["reload_me"])
```
interesting, why we need to change it to a list ?

br3ndonland replied:
Thanks for pointing this out. After reading through encode/uvicorn#978, I understand this better, and I reverted the change here.

If the user passes in a string, they can only pass in a single directory though, right? `Config(reload_dirs="reload_me, reload_me_too")` would result in `["reload_me, reload_me_too"]`.

euri10 replied:
yes, with click we have reload-dir that can be set multiple times, in which case click will set it as a tuple:
`uvicorn app:app --reload --reload-dir=dir1 --reload-dir=dir2` => `reload_dirs = ('dir1', 'dir2')` which is then passed onto Config
if it's just set one time as a string then we'll force it to be a list in the Config init, should be a tuple now but cant bother

euri10 commented on uvicorn/config.py:
```diff
@@ -107,8 +109,14 @@
 def create_ssl_context(
-    certfile, keyfile, password, ssl_version, cert_reqs, ca_certs, ciphers
-):
+    certfile: Union[Path, str],
```
why using different types here and below than those provided in the typeshed ?
```suggestion
    certfile: StrPath,
```

br3ndonland replied:
`StrPath` is in `_typeshed`, and is not available at runtime by default, so attempting to annotate with `StrPath` may raise `error: Name 'StrPath' is not defined [name-defined]`.

I'm happy to consider it, though. How would I import and use `StrPath`?

euri10 replied:
something like this ?
```
def create_ssl_context(
    certfile: Union[str, os.PathLike[str]],
    keyfile: Optional[Union[str, os.PathLike[str]]],
    password: Optional[str],
    ssl_version: int,
    cert_reqs: int,
    ca_certs: Optional[Union[str, os.PathLike[str]]],
    ciphers: Optional[str],
```

br3ndonland replied:
Sure, that works -> a2d9085. It's not subscriptable, so just has to be `os.PathLike` instead of `os.PathLike[str]`.

euri10 commented on uvicorn/config.py:
```diff
@@ -123,45 +131,45 @@ def create_ssl_context(
+        ws_max_size: int = 16 * 1024 * 1024,
+        lifespan: str = "auto",
+        env_file: Optional[Union[Path, str]] = None,
+        log_config: Optional[Union[dict, str]] = LOGGING_CONFIG,
```
can be a file too

br3ndonland replied:
Yes, but the file path would be coming in as a string, right? The log config logic assumes string in some places, like where `.endswith()` is used:

https://github.com/encode/uvicorn/blob/87da6cf4082cd28306bdb78d7d42552d131cc179/uvicorn/config.py#L252-L259

euri10 replied:
yes you're right, I had in minc click.Path could give you a Path object but it's not the case

euri10 commented on uvicorn/config.py:
```diff
@@ -300,22 +313,27 @@ def load(self):
+            if isinstance(self.app, str):
+                self.loaded_app = import_from_string(self.app)
+            else:
+                self.loaded_app = self.app
```
why do we do that ? afair it's done already in the import_from_string def, isn't it ?

br3ndonland replied:
Yeah, good point, we can just let `import_from_str` handle this. Reverted.

euri10 commented:
/rant on
this Approve and run button is really annoying
/rant off

br3ndonland committed:
Use os.PathLike for paths in uvicorn/config.py

https://github.com/encode/uvicorn/pull/1067#discussion_r646370076

Alternative to `pathlib.Path` introduced in Python 3.6.

br3ndonland commented:
> /rant on
> this Approve and run button is really annoying
> /rant off

Hopefully almost done 🤞

euri10 commented on tests/test_config.py:
```diff
@@ -15,34 +26,36 @@
 
 
-def wsgi_app(environ, start_response):
+def wsgi_app(environ: Any, start_response: Any) -> None:
```
I've made a draft PR that touched a little bit on wsgi app and came up with
```
Environ = typing.MutableMapping[str, typing.Any]
ExcInfo = typing.Tuple[    typing.Type[BaseException], BaseException, typing.Optional[TracebackType]]
StartResponse = typing.Callable[[str, typing.Iterable[typing.Tuple[str, str]], typing.Optional[ExcInfo]], None]
```

```suggestion
def wsgi_app(environ: Environ, start_response: StartResponse) -> None:
```

br3ndonland replied:
These are good suggestions. For the `Environ` type, I would modify it slightly, either simply `typing.MutableMapping` (which is what Starlette uses in [starlette/config.py](https://github.com/encode/starlette/blob/b6f3578bb2cf6c60e3efe110143409b47f368d36/starlette/config.py#L16)), or `typing.MutableMapping[str, str]`, based on [`os.environ`](https://github.com/python/cpython/blob/3e1c7167d86a2a928cdcb659094aa10bb5550c4c/Lib/os.py#L737) (which appears to expect all keys and values to be strings).

I went with `typing.MutableMapping`, hopefully that's cool.

euri10 replied:
the starlette Environ you're referring to has nothing to do with the Environ here I think, appart from the name, the latter is for configuring starlette iirc, the former is the dict for the wsgi spec (https://www.python.org/dev/peps/pep-0333/#environ-variables)
I think `typing.MutableMapping[str, str]` is incorrect since one of the values it can take are bytes, for the wsgi.input key

br3ndonland replied:
Ok that's fine, I accepted your suggestion (6e1c1dc).

euri10 commented on tests/test_config.py:
```diff
@@ -257,20 +286,25 @@ def test_log_config_file(mocked_logging_config_module):
 
 @pytest.fixture(params=["127.0.0.1", "127.0.0.2"])
-def forwarded_allow_ips(request):
+def forwarded_allow_ips(request: Any) -> Iterator[str]:
```
not sure about this, worth checking if Any could be avoided
```suggestion
def forwarded_allow_ips(request: FixtureRequest) -> Iterator[str]:
```

br3ndonland replied:
I agree, it's good to be more specific than `Any`. `pytest.FixtureRequest` works. The `param` attribute is [optional](https://docs.pytest.org/en/latest/reference/reference.html#pytest.FixtureRequest), so mypy requires a check for the attribute before indexing into it, like `getattr(request, "param")`.

euri10 reviewed (commented):
another round of comments, sorry this is a massive PR so I was not able to touch on everything at first sight :)
but we're getting there !

br3ndonland committed:
Use more specific types in test_config.py

https://github.com/encode/uvicorn/pull/1067#discussion_r647149143
https://github.com/encode/uvicorn/pull/1067#discussion_r647150799

https://github.com/encode/starlette/blob/b6f3578bb2cf6c60e3efe110143409b47f368d36/starlette/config.py#L16
https://github.com/python/cpython/blob/3e1c7167d86a2a928cdcb659094aa10bb5550c4c/Lib/os.py#L737
https://docs.pytest.org/en/latest/reference/reference.html#pytest.FixtureRequest

- Add custom types for exceptions and start response
- Use `MutableMapping` for `os.environ`: matches starlette/config.py.
- Use `pytest.FixtureRequest` for fixture requests. The `param`
  attribute is optional, so mypy requires a check for the attribute
  before indexing into it (`getattr(request, "param")`).

br3ndonland committed:
Install missing YAML type stubs for mypy

encode/uvicorn#1067

Fixes `[import]` error: Library stubs not installed for "yaml"

br3ndonland commented on requirements.txt:
```diff
@@ -15,6 +15,7 @@ isort
 pytest
 pytest-mock
 mypy
+types-pyyaml
```
mypy is complaining about the yaml imports now in GitHub Actions:

```text
mypy --show-error-codes

uvicorn/config.py:21: error: Library stubs not installed for "yaml" (or incompatible with Python 3.8)  [import]
uvicorn/config.py:21: note: Hint: "python3 -m pip install types-PyYAML"
uvicorn/config.py:21: note: (or run "mypy --install-types" to install all missing stub packages)
uvicorn/config.py:21: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
tests/test_config.py:18: error: Library stubs not installed for "yaml" (or incompatible with Python 3.8)  [import]
```

Running `mypy --install-types` didn't seem to fix it (in GitHub Actions at least), so I installed the `types-pyyaml` package from the typeshed. Otherwise, we would have to add ignores, like `import yaml # type: ignore[import]`.

Kludex replied:
Mypy 0.900 was released yesterday, and we don't pin the mypy version, jfyk.
https://mypy-lang.blogspot.com/2021/06/mypy-0900-released.html

EDIT: You have the message on the GitHub Actions and not locally because you need to upgrade it.

Kludex replied:
Can you try `mypy --install-types` instead of adding `types-pyyaml`?

euri10 replied:
kind of happy this doesnt blow up more than that !

br3ndonland commented:
> another round of comments, sorry this is a massive PR so I was not able to touch on everything at first sight :)
> but we're getting there !

Cool, thanks for your reviews! Good feedback.

Kludex commented on tests/test_config.py:
```diff
@@ -166,8 +188,10 @@ def test_ssl_config_combined(tls_certificate_pem_path):
 
-def asgi2_app(scope):
-    async def asgi(receive, send):  # pragma: nocover
+def asgi2_app(scope: Scope) -> Callable:
```
```suggestion
def asgi2_app(scope: Scope) -> Callable[[ASGIReceiveCallable, ASGISendCallable], Awaitable[None]]:
```

br3ndonland replied:
Could we use the `asgiref` types instead?

```py
def asgi2_app(scope: Scope) -> ASGI2Application:
    async def asgi(
        receive: ASGIReceiveCallable, send: ASGISendCallable
    ) -> None:  # pragma: nocover
        pass

    return asgi  # pragma: nocover

```

Kludex replied:
That doesn't work because the `ASGI2Application` is _all of it_. That part is only the `callable` that I've annotated.

br3ndonland replied:
> That doesn't work because the `ASGI2Application` is _all of it_. That part is only the `callable` that I've annotated.

I see. This is related to https://github.com/encode/uvicorn/pull/1067#discussion_r648638750 - if we don't want `Config(app)` to accept a callable, we might have to refactor these test methods.

Kludex commented on tests/test_config.py:
```diff
@@ -284,7 +322,7 @@ def test_env_file(web_concurrency: int, forwarded_allow_ips: str, caplog, tmp_pa
         config.load()
 
-    assert config.workers == int(os.getenv("WEB_CONCURRENCY"))
+    assert config.workers == int(str(os.getenv("WEB_CONCURRENCY")))
```
Why this is needed? Isn't the return already a `str`?

br3ndonland replied:
`config.workers` is an `int`, so in order to compare, the environment variable should also be an `int`.

The return type of `os.getenv()` is `Optional[str]` (if the specified environment variable is not set, the function call will return `None`). Attempting to run `int(None)` will raise a `TypeError`:

```text
>>> int(None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'
```

For type checking, either casting the object to a string first, or adding conditions like `if os.getenv("WEB_CONCURRENCY")` or `isinstance(os.getenv("WEB_CONCURRENCY"), str)` will work.

Kludex commented on uvicorn/config.py:
```diff
@@ -123,45 +130,45 @@ def create_ssl_context(
-        ssl_ciphers="TLSv1",
-        headers=None,
-        factory=False,
+        app: Union[ASGIApplication, Awaitable, Callable, str],
```
Why `Awaitable` and `Callable` should be accepted here?

br3ndonland replied:
The tests pass callables in:

https://github.com/encode/uvicorn/blob/a8af0da26c921b27849701f2cedbf89b8f3cbd86/tests/test_config.py#L37-L47

It might help to set up some proper ASGI application fixtures, instead of these simple callables.

Kludex replied:
I guess it only passes because `asgi_app` is not annotated, but if we annotate, probably `Union[ASGIApplication, str]` is enough. :thinking:

br3ndonland replied:
Well, what I'm saying is that if I change _uvicorn/config.py_ like this:

```py
class Config:
    def __init__(
        self,
        app: Union[ASGIApplication, str],
        ...
```

mypy raises errors on _tests/test_config.py_, like this:

```text
tests/test_config.py:97: error: Argument "app" to "Config" has incompatible type "Callable[[MutableMapping[str, Any], Callable[[str, Iterable[Tuple[str, str]], Optional[Tuple[Type[BaseException], BaseException, Optional[TracebackType]]]], None]], None]"; expected "Union[Union[Type[ASGI2Protocol], Callable[[Union[HTTPScope, WebsocketScope, LifespanScope], Callable[[], Awaitable[Union[HTTPRequestEvent, HTTPDisconnectEvent, WebsocketConnectEvent, WebsocketReceiveEvent, WebsocketDisconnectEvent, LifespanStartupEvent, LifespanShutdownEvent]]], Callable[[Union[HTTPResponseStartEvent, HTTPResponseBodyEvent, HTTPServerPushEvent, HTTPDisconnectEvent, WebsocketAcceptEvent, WebsocketSendEvent, WebsocketResponseStartEvent, WebsocketResponseBodyEvent, WebsocketCloseEvent, LifespanStartupCompleteEvent, LifespanStartupFailedEvent, LifespanShutdownCompleteEvent, LifespanShutdownFailedEvent]], Awaitable[None]]], Awaitable[None]]], str]"
tests/test_config.py:138: error: Argument "app" to "Config" has incompatible type "Callable[[], Union[Type[ASGI2Protocol], Callable[[Union[HTTPScope, WebsocketScope, LifespanScope], Callable[[], Awaitable[Union[HTTPRequestEvent, HTTPDisconnectEvent, WebsocketConnectEvent, WebsocketReceiveEvent, WebsocketDisconnectEvent, LifespanStartupEvent, LifespanShutdownEvent]]], Callable[[Union[HTTPResponseStartEvent, HTTPResponseBodyEvent, HTTPServerPushEvent, HTTPDisconnectEvent, WebsocketAcceptEvent, WebsocketSendEvent, WebsocketResponseStartEvent, WebsocketResponseBodyEvent, WebsocketCloseEvent, LifespanStartupCompleteEvent, LifespanStartupFailedEvent, LifespanShutdownCompleteEvent, LifespanShutdownFailedEvent]], Awaitable[None]]], Awaitable[None]]]]"; expected "Union[Union[Type[ASGI2Protocol], Callable[[Union[HTTPScope, WebsocketScope, LifespanScope], Callable[[], Awaitable[Union[HTTPRequestEvent, HTTPDisconnectEvent, WebsocketConnectEvent, WebsocketReceiveEvent, WebsocketDisconnectEvent, LifespanStartupEvent, LifespanShutdownEvent]]], Callable[[Union[HTTPResponseStartEvent, HTTPResponseBodyEvent, HTTPServerPushEvent, HTTPDisconnectEvent, WebsocketAcceptEvent, WebsocketSendEvent, WebsocketResponseStartEvent, WebsocketResponseBodyEvent, WebsocketCloseEvent, LifespanStartupCompleteEvent, LifespanStartupFailedEvent, LifespanShutdownCompleteEvent, LifespanShutdownFailedEvent]], Awaitable[None]]], Awaitable[None]]], str]"
tests/test_config.py:145: error: Argument "app" to "Config" has incompatible type "Callable[[], Union[Type[ASGI2Protocol], Callable[[Union[HTTPScope, WebsocketScope, LifespanScope], Callable[[], Awaitable[Union[HTTPRequestEvent, HTTPDisconnectEvent, WebsocketConnectEvent, WebsocketReceiveEvent, WebsocketDisconnectEvent, LifespanStartupEvent, LifespanShutdownEvent]]], Callable[[Union[HTTPResponseStartEvent, HTTPResponseBodyEvent, HTTPServerPushEvent, HTTPDisconnectEvent, WebsocketAcceptEvent, WebsocketSendEvent, WebsocketResponseStartEvent, WebsocketResponseBodyEvent, WebsocketCloseEvent, LifespanStartupCompleteEvent, LifespanStartupFailedEvent, LifespanShutdownCompleteEvent, LifespanShutdownFailedEvent]], Awaitable[None]]], Awaitable[None]]]]"; expected "Union[Union[Type[ASGI2Protocol], Callable[[Union[HTTPScope, WebsocketScope, LifespanScope], Callable[[], Awaitable[Union[HTTPRequestEvent, HTTPDisconnectEvent, WebsocketConnectEvent, WebsocketReceiveEvent, WebsocketDisconnectEvent, LifespanStartupEvent, LifespanShutdownEvent]]], Callable[[Union[HTTPResponseStartEvent, HTTPResponseBodyEvent, HTTPServerPushEvent, HTTPDisconnectEvent, WebsocketAcceptEvent, WebsocketSendEvent, WebsocketResponseStartEvent, WebsocketResponseBodyEvent, WebsocketCloseEvent, LifespanStartupCompleteEvent, LifespanStartupFailedEvent, LifespanShutdownCompleteEvent, LifespanShutdownFailedEvent]], Awaitable[None]]], Awaitable[None]]], str]"
tests/test_lifespan.py:199: error: Argument "app" to "Config" has incompatible type "Callable[[Any], Any]"; expected "Union[Union[Type[ASGI2Protocol], Callable[[Union[HTTPScope, WebsocketScope, LifespanScope], Callable[[], Awaitable[Union[HTTPRequestEvent, HTTPDisconnectEvent, WebsocketConnectEvent, WebsocketReceiveEvent, WebsocketDisconnectEvent, LifespanStartupEvent, LifespanShutdownEvent]]], Callable[[Union[HTTPResponseStartEvent, HTTPResponseBodyEvent, HTTPServerPushEvent, HTTPDisconnectEvent, WebsocketAcceptEvent, WebsocketSendEvent, WebsocketResponseStartEvent, WebsocketResponseBodyEvent, WebsocketCloseEvent, LifespanStartupCompleteEvent, LifespanStartupFailedEvent, LifespanShutdownCompleteEvent, LifespanShutdownFailedEvent]], Awaitable[None]]], Awaitable[None]]], str]"
Found 4 errors in 2 files (checked 32 source files)
```

I think some better test fixtures that actually instantiate ASGI applications would help. I'll look into this, but not sure how to do it right now.

Kludex replied:
I see. But it should not accept the Awaitable here. Can we remove it?

br3ndonland replied:
Done (2560310)

Kludex commented on uvicorn/config.py:
```diff
@@ -123,45 +130,45 @@ def create_ssl_context(
+        log_level: Optional[Union[str, int]] = None,
+        access_log: bool = True,
+        use_colors: Optional[bool] = None,
+        interface: Literal["auto", "asgi3", "asgi2", "wsgi"] = "auto",
```
I feel like this breaks your pattern (or you break it previously) comparing to:

```python
        http: Union[Type[asyncio.Protocol], str] = "auto",        
        ws: Union[Type[asyncio.Protocol], str] = "auto",
```
```python
        lifespan: str = "auto",
```
```python
        loop: str = "auto",
```
I mean, why don't make use of `Literal` on those as well?

br3ndonland replied:
Yes, good point. I set up some type aliases for these arguments (70694b5).

Kludex replied:
Cool! :tada:

Kludex commented on uvicorn/config.py:
```diff
@@ -360,12 +374,16 @@ def load(self):
+        if loop_setup_str:
+            loop_setup: Callable = import_from_string(loop_setup_str)
+            if not inspect.isfunction(loop_setup):
+                raise ImportFromStringError("Asyncio event loop must be a callable.")
```
@euri10 take special attention to this, plz
I'm not sure...

br3ndonland replied:
Yeah, I wasn't sure of the best approach here either. Open to suggestions.

euri10 replied:
I would type ignore that part

Kludex commented on uvicorn/importer.py:
```diff
@@ -1,13 +1,12 @@
 
 
-def import_from_string(import_str: Union[ModuleType, str]) -> ModuleType:
+def import_from_string(import_str: Any) -> Any:
```
I understand the current behavior as being wrong but is `Any` the best approach here?

br3ndonland replied:
This is one of the few places where I think `Any` is actually appropriate.

If any object type other than string is passed to `import_from_string`, it will just return that object. Otherwise, it will parse the string, import the module (and maybe the attribute after `:`), and return that module or attribute.

I tried setting up a type alias (https://github.com/encode/uvicorn/pull/1067#issuecomment-853388871), but it was really difficult to get it right.

Also, even if we get the typing correct within Uvicorn's source code, we don't know how other applications are using `import_from_string` at this point. `Any` would help avoid breaking type checking in other applications.

Kludex replied:
I see.

@euri10 Do you have something to say here or can we stick with `Any`?

euri10 replied:
no sounds about the best option with my limited mypy understanding,

Kludex commented:
Thanks for the PR @br3ndonland :)
I've left some comments, I hope you can take a look at them.

br3ndonland committed:
Merge branch 'master' of github.com:encode/uvicorn into mypy-config

br3ndonland committed:
Add Environ type to test_config.py

https://github.com/encode/uvicorn/pull/1067#discussion_r647149143

br3ndonland committed:
Add Literal type aliases for web server config

https://github.com/encode/uvicorn/pull/1067#discussion_r648498465

br3ndonland commented:
> Thanks for the PR @br3ndonland :)
> I've left some comments, I hope you can take a look at them.

Thank you for your help!

euri10 commented on tests/test_config.py:
```diff
@@ -125,27 +149,30 @@ def create_app():
     assert "--factory" in caplog.records[0].message
 
-    # App not a no-arguments callable.
+    # App not a no-arguments ASGIApplication.
```
I dont get this change, I'd leave it as is

Kludex replied:
Yeah, me too.

br3ndonland replied:
whoops, find and replace strikes again :zany_face:

euri10 commented on uvicorn/config.py:
```diff
@@ -31,46 +32,52 @@
 from uvicorn.middleware.wsgi import WSGIMiddleware
 
-TRACE_LOG_LEVEL = 5
+HttpProtocolType = Literal["auto", "h11", "httptools"]
```
```suggestion
HTTPProtocolType = Literal["auto", "h11", "httptools"]
```

euri10 commented on uvicorn/config.py:
```diff
@@ -31,46 +32,52 @@
 
-TRACE_LOG_LEVEL = 5
+HttpProtocolType = Literal["auto", "h11", "httptools"]
+WsProtocolType = Literal["auto", "none", "websockets", "wsproto"]
```
```suggestion
WSProtocolType = Literal["auto", "none", "websockets", "wsproto"]
```

euri10 commented on uvicorn/config.py:
```diff
@@ -280,11 +298,11 @@ def configure_logging(self):
 
-        if self.is_ssl:
-            self.ssl = create_ssl_context(
+        if self.is_ssl and self.ssl_certfile:
```
why do we need to add this ` and self.ssl_certfile` this seems odd, is_ssl check if we have a keyfile or certfile already ?

br3ndonland replied:
`certfile` is a required positional argument when running [`SSLContext.load_cert_chain`](https://docs.python.org/3/library/ssl.html#ssl.SSLContext.load_cert_chain).

`is_ssl` could return `True` if `certfile` is not present:

https://github.com/encode/uvicorn/blob/f025efd4a853e87921b07acdb6efb1d25931a686/uvicorn/config.py#L239-L242

euri10 replied:
ok so the fix is wrong imho since if someone forgets to pass a certfile like `uvicorn dev.app:app --ssl-keyfile=dev/server.key` then the server will start but without ssl, which is deceiving

I'd propose instead :
```
        if self.is_ssl:
            assert self.ssl_certfile
            self.ssl = create_ssl_context(
```
so that the server does not start on that config mistake and it solves the mypy error
@Kludex ?

br3ndonland replied:
I think the assertion would be reasonable here. I added it (8b38633), and I'm happy to consider alternatives also.

br3ndonland committed:
Use suggested casing for Literal type aliases

https://github.com/encode/uvicorn/pull/1067#discussion_r649148251
https://github.com/encode/uvicorn/pull/1067#discussion_r649148543

br3ndonland committed:
Restore test_config.py test_app_factory comment

https://github.com/encode/uvicorn/pull/1067#discussion_r649140031

Incorrectly modified in c436bba.

br3ndonland committed:
Simplify event loop setup in config.py

https://github.com/encode/uvicorn/pull/1067#discussion_r649153064

euri10 committed:
Merge branch 'master' into mypy-config

euri10 commented:
I forgot to remove the type def in line 325 when merging master, can you update ?

br3ndonland committed:
Remove old type comment after merging master

https://github.com/encode/uvicorn/pull/1067#issuecomment-859657665

br3ndonland committed:
Assert that certfile is present for SSL context

https://github.com/encode/uvicorn/pull/1067#discussion_r650065696
https://docs.python.org/3/library/ssl.html#ssl.SSLContext.load_cert_chain

`certfile` is a required argument for  `SSLContext.load_cert_chain`. As
it is currently, `is_ssl` could return `True` without `certfile`.

br3ndonland commented:
> I forgot to remove the type def in line 325 when merging master, can you update ?

Sure, no problem, I updated and the CI checks should be passing now.

br3ndonland commented:
Thanks @euri10 and @Kludex for all your help on this PR so far.

I will be AFK for a few days, but will check in on this when I return.

euri10 commented on tests/test_config.py:
```diff
@@ -2,47 +2,69 @@
+StartResponse = typing.Callable[
+    [str, typing.Iterable[typing.Tuple[str, str]], typing.Optional[ExcInfo]], None
+]
+
```
let's move this to _types.py so that we can use it for the wsgi middleware in the future, with a comment stating it's wsgi stuff

br3ndonland replied:
Good idea, we can continue adding custom types there. (f9f509b)

euri10 commented on uvicorn/config.py:
```diff
@@ -368,12 +389,11 @@ def load(self):
-            loop_setup()
+    def setup_event_loop(self) -> None:
+        loop_setup: Callable = import_from_string(LOOP_SETUPS[self.loop])
+        loop_setup()
```
why the `if loop_setup is not None:` disappears ?

br3ndonland replied:
Restored (d93bf88)

euri10 reviewed (commented):
ok 2 more comments on my side. I'm sorry this takes so long but it's a huge piece of review, hopefully we get to the end :)

euri10 commented:
> Thanks @euri10 and @Kludex for all your help on this PR so far.
> 
> I will be AFK for a few days, but will check in on this when I return.

hopefully you're back  @br3ndonland  ;) can we try to finish this, I know it took already a while but I feel we're near the end !

br3ndonland committed:
Restore support for Config(loop='none')

encode/uvicorn#455
encode/uvicorn@e9e59e9
encode/uvicorn@e60ba66
https://github.com/encode/uvicorn/pull/1067#discussion_r650393315

br3ndonland committed:
Move WSGI types to uvicorn/_types.py

https://github.com/encode/uvicorn/pull/1067#discussion_r650392993

br3ndonland commented:
> hopefully you're back @br3ndonland ;) can we try to finish this, I know it took already a while but I feel we're near the end !

I'm back! Yes I think we're almost done. Thanks for sticking with it!

euri10 reviewed (approved):
As far as I'm concerned I'm happy with it.
@Kludex can you please take another look and clear if necessary the conversation [here] (https://github.com/encode/uvicorn/pull/1067/files#r648495399) (will merge once I'm sure nothing's left unanswered!)
Thanks a lot for your time and patience with us on this @br3ndonland but this was not an easy one !!

Kludex commented:
Only the `Awaitable` in the `app` parameter on the `Config` and then we're ready! 🙌

br3ndonland committed:
Remove Awaitable from app type annotation

https://github.com/encode/uvicorn/pull/1067#issuecomment-864770712

br3ndonland commented:
> Only the `Awaitable` in the `app` parameter on the `Config` and then we're ready! 🙌

Done (2560310). I was thinking about removing `Awaitable`. Thanks for pointing it out.

In a future PR, I will also look into removing `Callable` from the `app` annotation, as discussed in https://github.com/encode/uvicorn/pull/1067#discussion_r648653118.

br3ndonland committed:
Merge branch 'master' of github.com:encode/uvicorn into mypy-config

br3ndonland commented:
Merged in latest `master`. Should be a clean PR merge with all CI checks passing.

euri10 commented:
let's roll, thanks a lot @br3ndonland !

Kludex commented:
Thanks @br3ndonland ! :)
